### ISSUE TITLE
How does --challenge-alias works internally to avoid concurrency problem ?

### ISSUE THREAD

**Issue by tristanbes** [Neilpang/acme.sh#1527]
Hello,

How does the challenge-alias works in case of multiple domains having the same `--challenge-alias` ? Are requests queued ? (One domain gets validated after one domain ?)

Is there anything preventing from running 2 instance of `acme.sh` at the same moment and then having problem with concurrency when using DNS validation mode with an alias ? 

Ex:
Shell 1: `acme.sh --issue -d '*.example.fr' --challenge-alias example-proxy.fr --dns dns_cf`

Shell 2, 1sec later: `acme.sh --issue -d '*.2example.com' --challenge-alias example-proxy.fr --dns dns_cf`

In this case, it would mean that 2 DNS record would be written/overwiten before the first one being validated right ?

So: is it up to us to ensure that only 1 `acme.sh` command is running at a time ? Or did you implement some kind of protection against this ?

Thank you.

**Comment by martgras** [comment]
It depends on the dns api implementation. If it is well written there won't be a problem. The dns provider should only add new values to the txt record instead of overriding existing ones.  Wait for validation and then remove only the values that were added before. 
I can only speak for my dns_azure plugin that works that way. However there is a chance of a concurrently problem that I'm going to address soon. 
I'm removing the txt record if the last value was removed to avoid an empty txt record.  If you start a second instance that has added the value after my code checks for existing values and before the record was removed your new value will be removed as well. 
Unlikely but not impossible..
So in short it depends. Any dnsapi that removes the complete record instead of just the validation value can run into this.
You might get an idea by looking at the dns_rm code of the provider you are using.

**Comment by tristanbes** [comment]
Ok thanks for your explanations guys !
